rapids-cmake/cpm/detail/pinning_write_file.cmake

```diff
@@ -0,0 +1,259 @@
+  set(_RAPIDS_URL)
+  set(_RAPIDS_SHA)
+  if(EXISTS "${source_dir}")
+    execute_process(COMMAND ${GIT_EXECUTABLE} ls-remote --get-url
+                    WORKING_DIRECTORY ${source_dir}
+                    ERROR_QUIET
+                    OUTPUT_VARIABLE _RAPIDS_URL
+                    OUTPUT_STRIP_TRAILING_WHITESPACE)
+    # Need to handle when we have applied N patch sets to the git repo and therefore the latest
+    # commit is just local
+    #
+    # Find all commits on our branch back to the common parent ( what we cloned )
+    #
+    execute_process(COMMAND ${GIT_EXECUTABLE} show-branch --current --sha1-name
+                    WORKING_DIRECTORY ${source_dir}
+                    ERROR_QUIET
+                    OUTPUT_VARIABLE _rapids_commit_stack
+                    OUTPUT_STRIP_TRAILING_WHITESPACE)
+    # The last entry in the output that has "* [" is our commit
+    #
+    # Find that line and convert the `* [short-sha1] Commit Message` to a list that is ` *
+    # ;short-sha1;Commit Message` and extract the short sha1
+    string(FIND "${_rapids_commit_stack}" "* [" position REVERSE)
+    if(position LESS 0)
+      # No changes to the repo so use the `HEAD` keyword
+      set(short_sha HEAD)
+    else()
+      string(SUBSTRING "${_rapids_commit_stack}" ${position} -1 _rapids_commit_stack)
+      string(REGEX REPLACE "(\\[|\\])" ";" _rapids_commit_stack "${_rapids_commit_stack}")
+      list(GET _rapids_commit_stack 1 short_sha)
+    endif()
+
+    # Convert from the short sha1 ( could be keyword `HEAD` ) to a full SHA1
+    execute_process(COMMAND ${GIT_EXECUTABLE} rev-parse ${short_sha}
+                    WORKING_DIRECTORY ${source_dir}
+                    ERROR_QUIET
+                    OUTPUT_VARIABLE _RAPIDS_SHA
+                    OUTPUT_STRIP_TRAILING_WHITESPACE)
+  endif()
+  # Only set the provided variables if we extracted the information
+  if(_RAPIDS_URL)
+    set(${git_url_var} "${_RAPIDS_URL}" PARENT_SCOPE)
+  endif()
+  if(_RAPIDS_SHA)
+    set(${git_sha_var} "${_RAPIDS_SHA}" PARENT_SCOPE)
+  endif()
+
+endfunction()
+
+#[=======================================================================[.rst:
+rapids_cpm_pinning_add_json_entry
+---------------------------------
+
+.. versionadded:: v24.04.00
+
+#]=======================================================================]
+function(rapids_cpm_pinning_add_json_entry json_var package_name url_var sha_var)
+  include("${rapids-cmake-dir}/cpm/detail/get_default_json.cmake")
+  include("${rapids-cmake-dir}/cpm/detail/get_override_json.cmake")
+  get_default_json(${package_name} json_data)
+  get_override_json(${package_name} override_json_data)
+
+  set(url_string)
+  set(sha_string)
+  if(${url_var})
+    string(CONFIGURE [=["git_url": "${${url_var}}",]=] url_string)
+  endif()
+  if(${sha_var})
+    string(CONFIGURE [=["git_tag": "${${sha_var}}",]=] sha_string)
+  endif()
+  # We start with a default template, and only add members that don't exist
```

Comment:
I assume this is mostly for setting things like always_download, patches, or proprietary binary. Could we also set the URL/SHA? If so, then maybe this relates to my previous comment regarding the url/sha failure modes. If those could be overwritten, then maybe we need to error-check the final output JSON rather than erroring if the variables weren't set up front to allow the delayed set.

Reply:
Due to projects like `nvcomp` we can't presume that the 'cloned repository` will have any git information.

So what we are doing here is constructing the set of json entries that are most correct. So by not adding `git_tag`, and `git_url` we can extract them from the override or `version.json` entry. This simplifies the following logic since we can just iterate the override and default entries adding any elements that don't exist in our new entry.

Reply:
That makes sense, so my main concern is really, is there a possibility that we could encoded a pinned version for a dependency that we later don't know how to acquire? I guess not, because if we were able to get it in the first place we must have done so somehow (be it via git, a download URL, or something else) and that same CMake logic is expected to be available at the time we use the pinning file, so there's no world where we get a pinning file with a dependency that we don't know how to acquire. Is that correct?

Reply:
> so there's no world where we get a pinning file with a dependency that we don't know how to acquire. Is that correct?

Correct!

Reply:
Looking at this more, I believe we need to handle the use case when `CPM` finds a local version of a package ( e.g. fmt ) instead of building from source. In those cases `CPM_PACKAGES` will still list `fmt`, but it won't have a source dir.

I will need to extend `rapids_cpm_pinning_write_file` to handle checking that `CPM` actually added a CMake `FetchContent` record for the project before we try and construct a JSON entry for it.

Reply:
Corrected this bug, added a test so we don't regress it.

Reply:
OK yes this kind of case was making me a bit uneasy. So the assumption in this case is that users of the pinned file will have the same version installed locally as whenever the pin file was created in order to fully reproduce the build. Should we include some information in the file to indicate that a dependency was found? How do we want to communicate this to users? For cases where complete reproducibility is required should we recommend setting `CPM_DOWNLOAD_ALL`?

Reply:
> So the assumption in this case is that users of the pinned file will have the same version installed locally as whenever the pin file was created in order to fully reproduce the build

Correct. The same assumption is applied for any existing `find_package` or `rapids_find_package` call as well. The end goal here is to pin packages we are building by source. 

In theory the creation of this file is done in a controlled env, like CI which has `CPM_DOWNLOAD_ALL` set so that we have a reproducible baseline env. From that point we just need to do the transformation from branches to exact SHA1. But that might not always be the case.

> Should we include some information in the file to indicate that a dependency was found? How do we want to communicate this to users? For cases where complete reproducibility is required should we recommend setting CPM_DOWNLOAD_ALL?

We currently don't have a way to represent an informative package in the json file. So I am more inclined to have the function output a message stating what packages it failed to record and advise people to also set `CPM_DOWNLOAD_ALL`.

I think that in a future PR we can extend the `version.json` to have the concept of `CPM_DOWNLOAD_ALL` which would allow for a declarative declaration

Reply:
OK cool I agree with that plan. I can approve once we have that messaging in place.